LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/Regions/Region.cs

```diff
@@ -241,9 +243,15 @@ public string GetDownstreamDR(Rxpk upstreamChannel)
             {
                 var dr = this.DRtoConfiguration.FirstOrDefault(x => x.Value.configuration == upstreamChannel.Datr).Key;
                 // TODO take care of rx1droffset
-                var (configuration, maxPyldSize) = this.DRtoConfiguration[10 - dr];
-                return configuration;
-                // return this.DRtoConfiguration[10-dr].configuration;
+                if (dr >= 0 && dr < 5)
+                {
+                    var (configuration, maxPyldSize) = dr != 4 ? this.DRtoConfiguration[10 + dr] : this.DRtoConfiguration[13];
+                    return configuration;
+                }
+                else
+                {
+                    throw new Exception("Datarate in {0} region was not within the acceptable range of upstream datarates.");

```

Comment:
You forgot to supply the value for {0}. I think $"" are more readable than string.Format("{0}"). Perhaps add the input data rate too so that the exception message will give us all information needed to understand the problem (region, data rate).

Reply:
Good one, thank you!